src/main/scala/io/iohk/ethereum/ledger/BlockImport.scala

```diff
@@ -0,0 +1,264 @@
+    */
+  private def executeBlocks(blocks: List[Block], parentTd: BigInt): (List[BlockData], Option[BlockExecutionError]) = {
+    @tailrec
+    def go(executedBlocks: List[BlockData], remainingBlocks: List[Block], parentTd: BigInt, error: Option[BlockExecutionError])
+    :(List[BlockData], Option[BlockExecutionError]) ={
+      if (remainingBlocks.isEmpty) {
+        (executedBlocks.reverse, None)
+      } else if (error.isDefined) {
+        (executedBlocks, error)
+      } else {
+        val blockToExecute = remainingBlocks.head
+        executeBlock(blockToExecute, true) match {
+          case Right (receipts) =>
+            val td = parentTd + blockToExecute.header.difficulty
+            val newBlockData = BlockData(blockToExecute, receipts, td)
+            blockchain.save(newBlockData.block, newBlockData.receipts, newBlockData.td, saveAsBestBlock = true)
+            go(newBlockData :: executedBlocks, remainingBlocks.tail, td, None)
+          case Left(executionError) =>
+            go(executedBlocks, remainingBlocks, 0, Some(executionError))
+        }
+      }
+    }
+
+    go(List.empty[BlockData], blocks, parentTd, None)
+  }
+
+  private def handleImportTopValidationError(
+    error: ValidationBeforeExecError,
+    block: Block,
+    bestBlockBeforeImport: Block,
+    blockImportResult: BlockImportResult
+  ): BlockImportResult = {
+    blockImportResult match {
+      case BlockImportedToTop(blockImportData) =>
+        blockImportData.foreach { blockData =>
+          val hash = blockData.block.header.hash
+          blockQueue.removeSubtree(hash)
+          blockchain.removeBlock(hash, withState = true)
+        }
+        blockchain.saveBestKnownBlock(bestBlockBeforeImport.header.number)
+      case _ => ()
+    }
+    handleBlockValidationError(error, block)
+  }
+
+  private def handleBlockValidationError(error: ValidationBeforeExecError, block: Block): BlockImportResult = error match {
+    case ValidationBeforeExecError(HeaderParentNotFoundError) =>
+      log.debug(s"Block(${block.idTag}) has unknown parent")
+      UnknownParent
+
+    case ValidationBeforeExecError(reason) =>
+      log.debug(s"Block(${block.idTag}) failed pre-import validation")
+      BlockImportFailed(reason.toString)
+  }
+
+  private[ledger] def reorganise(block: Block, currentBestBlock: Block, currentTd: BigInt): Future[BlockImportResult] = Future {
+    blockValidation.validateBlockBeforeExecution(block).fold(error => handleBlockValidationError(error, block), _ => {
+      blockQueue.enqueueBlock(block, currentBestBlock.header.number) match {
+        case Some(Leaf(leafHash, leafTd)) if isBetterBranch(block, currentBestBlock, leafTd, currentTd) =>
+          log.debug("Found a better chain, about to reorganise")
+          reorganiseChain(leafHash, leafTd)
+
+        case _ =>
+          BlockEnqueued
+      }
+    })
+  }(validationContext)
```

Comment:
In old version `reorganise` was taking `blockexecutioncontext` as implicit, and wa using it. I would like to keep it that way. I mean the callet would suprised that now all his execution is happening on validation context.